retroshare-gui/src/gui/chat/ChatWidget.cpp

```diff
@@ -393,30 +430,99 @@ void ChatWidget::init(const ChatId &chat_id, const QString &title)
 				if (time(nullptr) <= historyIt->recvTime+2)
 					continue;
 
-				QString name;
-				if (chatId.isLobbyId() || chatId.isDistantChatId())
-				{
-					RsIdentityDetails details;
-					time_t start = time(nullptr);
-					while (!rsIdentity->getIdDetails(RsGxsId(historyIt->peerName), details))
-					{
-						std::this_thread::sleep_for(std::chrono::milliseconds(10));
-						if (time(nullptr)>start+2)
-						{
-							std::cerr << "ChatWidget History haven't found Id Details and have wait 1 sec for it." << std::endl;
-							break;
+				const std::string
+					notbeinghashableisagoodidea((const char*)historyIt->chatPeerId.toByteArray());
+				struct lookup_info& state = statecache[notbeinghashableisagoodidea];
+				if(!state.hasName) {
+					int tries;
+					for(tries=0;tries<4;++tries) {
+						assert(!state.hasName);
+						if(state.isDirect) {
+						GOT_DIRECT:
+							{
+								// direct chat, with no RxGxsId, only an RsPeerId
+								RsPeerDetails info;
+								if(rsPeers->getPeerDetails(state.peer, info)) {
+									/*
+									 * info.name is the profile name NOT the node name
+									 * 
+									 * info.location is the node name,  as set in checkAccount, in
+									 * libretroshare/src/rsserver/rsaccounts.cc
+									 */
+									state.name = info.name;
+									state.hasName = true;
+									break;
+								}
+							}
+						} else if(state.hasGxs) {
+						GOT_GXS:
+							{
+								// we need the RsGxsId name still though
+								RsIdentityDetails details;
+								if(rsIdentity->getIdDetails(state.gxs, details)) {
+									state.name = details.mNickname;
+									state.hasName = true;
+									break;
+								}
+							}
+						} else if(chatId.isLobbyId()) {
+							// XXX:
+							// no joke, this is how libretroshare/src/pqi/p3historymgr.c ACTUALLY
```

Comment:
you're right. It's quite bad because it supposes that the endian will be kept between two calls. Since history is loaded and saved on the same machine, it is the case. however, that conversion is inelegant for the least.